akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -381,6 +440,53 @@ final class CoordinatedShutdown private[akka] (
    */
   private[akka] def jvmHooksLatch: CountDownLatch = _jvmHooksLatch.get
 
+  /**
+   * Scala API: Add a task to a phase, returning an object which will cancel it
+   * on demand and remove it from the task pool (so long as the same task has not
+   * been added elsewhere). Tasks in a phase are run concurrently, with no ordering
+   * assumed.
+   *
+   * Adding a task to a phase does not remove any other tasks from the phase.
+   *
+   * If the same task is added multiple times, each addition must be canceled
+   * in order to remove the task from the phase. Such a task is in any case not
+   * run more than once.
```

Comment:
Why the de-duplication? Wouldn't that also be a subtle change of behavior for the old addTask?

also strange that "each addition must be canceled" but only one is run.

Reply:
So this comes out of using thunks as keys in a hashmap (which uses their default `.hashcode()` behavior, which I'm pretty sure is `System.identityHashCode()` unless the user is passing in an overriding class). In the current implementation, if the same thunk is registered twice, it shows up twice in the vector (and will get run twice).

This may be a failure of my imagination, but I can't come up with a use case where a particular concrete thunk is passed around and registered multiple times and the users expect it to be *run* multiple times concurrently. However, I can absolutely see a usage where the passed-in thunk is a method of a class, which can be registered from several code paths; that code running multiple times concurrently on shutdown could very easily be a source of bugs.

So yes, this is a change of (undocumented and untested) behavior, and I believe the new behavior is more in line with what users would expect out of the box.

Reply:
What do you base the assumption that the de-duplication based on equals of the thunk is "what users would expect"? Registering twice, running twice is less magic, more expected, in my opinion.

That is how other libraries (that I know of) and also other places within Akka behaves for callbacks like this. For example `Future.onComplete`.

Reply:
OK, I can buy that. This will also make the updates easier to read, I think.